features/exploredash/src/main/java/org/dash/wallet/features/exploredash/ExploreDatabase.kt

```diff
@@ -0,0 +1,190 @@
+    Merchant::class,
+    MerchantFTS::class,
+    Atm::class,
+    AtmFTS::class
+], version = 1)
+@TypeConverters(RoomConverters::class)
+abstract class ExploreDatabase : RoomDatabase() {
+    abstract fun merchantDao(): MerchantDao
+    abstract fun atmDao(): AtmDao
+
+    companion object {
+        private val log = LoggerFactory.getLogger(ExploreDatabase::class.java)
+        private var instance: ExploreDatabase? = null
+
+        fun getAppDatabase(context: Context, config: Configuration): ExploreDatabase {
+            if (instance == null) {
+                instance = open(context, config)
+            }
+            return instance!!
+        }
+
+        suspend fun updateDatabase(
+            context: Context,
+            config: Configuration,
+            repository: ExploreRepository
+        ) {
+            log.info("force update explore db")
+            if (instance != null) {
+                instance!!.close()
+            }
+            instance = update(context, config, repository)
+        }
+
+        private fun open(context: Context, config: Configuration): ExploreDatabase {
+            val dbBuilder = Room.databaseBuilder(
+                context,
+                ExploreDatabase::class.java,
+                config.exploreDatabaseName
+            )
+
+            log.info("Open database {}", config.exploreDatabaseName)
+            return dbBuilder
+                .fallbackToDestructiveMigration()
+                .build()
+        }
+
+        private suspend fun update(
+            context: Context,
+            config: Configuration,
+            repository: ExploreRepository
+        ): ExploreDatabase {
+            val dbUpdateFile = repository.getUpdateFile()
+            var exploreDatabaseName = config.exploreDatabaseName
+
+            if (dbUpdateFile.exists()) {
+                val dbTimestamp = repository.getTimestamp(dbUpdateFile)
+                log.info(
+                    "found explore db update package {}, with a timestamp: {}",
+                    dbUpdateFile.absolutePath,
+                    dbTimestamp
+                )
+                val oldDbFile = context.getDatabasePath(exploreDatabaseName)
+                repository.markDbForDeletion(oldDbFile)
+                exploreDatabaseName = config.setExploreDatabaseName(dbTimestamp)
+            }
+
+            val dbBuilder = Room.databaseBuilder(
+                context,
+                ExploreDatabase::class.java,
+                exploreDatabaseName
+            )
+
+            return preloadAndOpen(dbBuilder, repository, dbUpdateFile)
+        }
+
+        @VisibleForTesting
+        suspend fun preloadAndOpen(
+            dbBuilder: Builder<ExploreDatabase>,
+            repository: ExploreRepository,
+            dbUpdateFile: File
+        ): ExploreDatabase {
+            require(dbUpdateFile.exists()) { "dbUpdateFile doesn't exist"}
+
+            return suspendCancellableCoroutine { coroutine ->
```

Comment:
The worker needs to wait for the preload operation to ensure that `localDatabaseTimestamp` is updated; to achieve it this method is turned into a suspend method.